src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+    }
+    stateHistory.push(currentState);
+  });
+
+  return stateHistory;
```

Comment:
[CODE STYLE] - The object copy should be named properly to enhance readability. Instead of `currentState`, a more descriptive name like `stateClone` or `stateCopy` would be better.